AuthN: Optionally use tokens for unified storage client authentication

drclau opened the pull request:
Part of https://github.com/grafana/identity-access-team/issues/589

This PR introduces authentication based on ID and Access tokens to the Unified Storage client.
The server side PR implementation has been done in #95086

The implementation makes use of the functionality provided by [authlib](https://github.com/grafana/authlib). On both client and server-side 3 setups are supported:
- **in-proc:** uses gRPC over Go channels.
  - **Note:** `in-proc mode` has been implemented in a [separate PR](https://github.com/grafana/grafana/pull/93124)
- **gRPC on-prem:** uses gRPC over TCP sockets. In contrast to cloud mode, on-prem has Access tokens disabled.
- **gRPC cloud:** this is the full implementation.

To allow for a smooth transition, the following steps have been taken:
- client-side: the default is to use the old mechanism, with the new one being enabled via featureflag
- server-side: we handle both the new authentication flow and the current one.

**Diagrams:**

**in-proc mode:**
![image](https://github.com/user-attachments/assets/59732ff7-5fda-4201-aebc-a431f602a16d)

**on-prem gRPC mode:**
![image](https://github.com/user-attachments/assets/2378e629-701c-4681-8256-c3f9f62c254d)

**Cloud gRPC mode:**
![image](https://github.com/user-attachments/assets/fb961174-ed2b-4cc4-ba29-842f79696bd4)

drclau committed:
authn: client side updates

Co-Authored-By: Gabriel MABILLE <gamab@users.noreply.github.com>

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
wip

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
allow insecure connections (for testing purposes); remove audience checks

audience checks will still need to be done for Access tokens, but not for ID tokens

drclau committed:
allow insecure conns in dev mode + refactoring

ryantxu commented on pkg/storage/unified/sql/service.go:
I know this is early -- but nothing auth related should be in the sql package.   It should live in the `storage/unified/resource` package

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
finalize authlib grpc interceptors usage

drclau committed:
make update-workspace

drclau committed:
cleanup: remove unused `GrpcServerConfig.Mode`

drclau committed:
update authlib version (small fix)

drclau committed:
cleanup: refactoring leftover

drclau committed:
make update-workspace

gamab commented on pkg/services/authn/grpcutils/config.go:
I wonder if and how we could support flags for this (later PR)

gamab commented on pkg/services/authn/grpcutils/config.go:
```diff
@@ -0,0 +1,35 @@
+	AllowedAudiences []string
+}
+
+func ReadGprcServerConfig(cfg *setting.Cfg) (*GrpcServerConfig, error) {
```
No error needed anymore.
```suggestion
func ReadGprcServerConfig(cfg *setting.Cfg) (*GrpcServerConfig) {
```

DanCech replied:
error is needed, but function name has a typo:
```suggestion
func ReadGrpcServerConfig(cfg *setting.Cfg) (*GrpcServerConfig, error) {
```

gamab replied:
You are write, seems like we reintroduces modes :+1:

gamab commented on pkg/services/authn/grpcutils/grpc_authenticator.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func NewInProcGrpcAuthenticator() *authnlib.GrpcAuthenticator {
+	return authnlib.NewUnsafeGrpcAuthenticator(
```
```suggestion
	// In proc grpc ID token signature verification can be skipped
	return authnlib.NewUnsafeGrpcAuthenticator(
```

gamab commented on pkg/services/authz/client.go:
```diff
@@ -173,3 +176,16 @@ func newCloudLegacyClient(authCfg *Cfg) (authzlib.MultiTenantClient, error) {
+		grpc.WithUnaryInterceptor(interceptor.UnaryClientInterceptor),
+		grpc.WithStreamInterceptor(interceptor.StreamClientInterceptor),
+	}
+	if allowInsecure {
```
:+1:

gamab commented on pkg/services/authz/client.go:
```diff
@@ -111,27 +112,32 @@ func newInProcLegacyClient(server *legacyServer) (authzlib.MultiTenantClient, er
 	grpcClientConfig := authnlib.GrpcClientConfig{}
 	clientInterceptor, err := authnlib.NewGrpcClientInterceptor(&grpcClientConfig,
 		authnlib.WithDisableAccessTokenOption(),
+		authnlib.WithIDTokenExtractorOption(func(ctx context.Context) (string, error) {
```
If I'm not mistaken, we don't need ID tokens in this call.

gamab commented on pkg/services/authz/config.go:
```diff
@@ -55,5 +57,6 @@ func ReadCfg(cfg *setting.Cfg) (*Cfg, error) {
 		token:            token,
 		tokenExchangeURL: tokenExchangeURL,
 		tokenNamespace:   tokenNamespace,
+		allowInsecure:    cfg.Env == "development",
```
:+1:

gamab replied:
```suggestion
		allowInsecure:    cfg.Env == setting.Dev,
```

gamab commented on pkg/storage/unified/sql/service.go:
```diff
@@ -62,7 +62,10 @@ func ProvideService(
 	}
 
-	authn := &grpc.Authenticator{}
+	authn, err := grpcutils.NewGrpcAuthenticator(cfg)
```
Nice :100:

gamab commented on pkg/services/grpcserver/service.go:
```diff
@@ -71,19 +73,30 @@ func ProvideService(cfg *setting.Cfg, features featuremgmt.FeatureToggles, authe
 	var opts []grpc.ServerOption
 
+	// We don't need to support the CloudNamespaceFormatter here, because the grpcserver is only used on-prem.
+	// TODO(drclau): validate this assumption.
```
As discussed with you, on second thought we need to support cloud

gamab commented on pkg/services/grpcserver/service.go:
```diff
@@ -71,19 +73,30 @@ func ProvideService(cfg *setting.Cfg, features featuremgmt.FeatureToggles, authe
+		authzlib.WithDisableAccessTokenNamespaceAccessCheckerOption(),
+		authzlib.WithIDTokenNamespaceAccessCheckerOption(true),
+	)
+	stackIdExtractor := authzlib.MetadataStackIDExtractor(authzlib.DefaultStackIDMetadataKey)
```
Self note: Verify client side

drclau committed:
authz client cfg changes

- removed ModeCloud, relying on ModeGrpc and stackID instead to discover if we're running in Cloud
- reusing settings from "grpc_client_authentication", instead of duplicating in "authorization" section

Co-Authored-By: Gabriel MABILLE <gamab@users.noreply.github.com>

drclau committed:
added stackIdExtractor for cloud mode

Co-Authored-By: Gabriel MABILLE <gamab@users.noreply.github.com>

drclau committed:
refactoring NamespaceAccessChecker usage and use CloudNamespaceFormatter in Cloud

Co-Authored-By: Gabriel MABILLE <gamab@users.noreply.github.com>

drclau committed:
Update pkg/services/authn/grpcutils/grpc_authenticator.go

Co-authored-by: Gabriel MABILLE <gamab@users.noreply.github.com>

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
Update authlib

gamab committed:
Use NamespaceAuthorizer

gamab committed:
Add Auth func overrides

gamab committed:
Mod update

gamab committed:
Else statement

drclau committed:
removed WithIDTokenExtractorOption, and other PR feedback

gamab committed:
Fix trace

drclau committed:
make update-workspace

drclau committed:
update usage of ReadGprcServerConfig()

gamab commented on pkg/storage/unified/resource/client_wrapper.go:
```diff
@@ -1,30 +1,145 @@
+				conn,
+				clientInterceptor.UnaryClientInterceptor,
+				clientInterceptor.StreamClientInterceptor,
+			)),
+		nil
```
```suggestion
			)), nil
```

drclau committed:
refactoring (aesthetics)

Co-Authored-By: Gabriel MABILLE <gamab@users.noreply.github.com>

drclau committed:
refactoring

Co-Authored-By: Gabriel MABILLE <gamab@users.noreply.github.com>

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

ryantxu commented on go.mod:
any intentional changes here?

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
Align

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
Rename NewResourceClient funcs

gamab commented on pkg/storage/unified/resource/client.go:
```diff
@@ -1,54 +1,129 @@
+		return nil, err
+	}
+
+	return NewResourceStoreClient(grpchan.InterceptClientConn(conn, clientInt.UnaryClientInterceptor, clientInt.StreamClientInterceptor)), nil
```
```suggestion
	conn, err := grpc.NewClient(*addr, grpc.WithTransportCredentials(creds), grpc.WithUnaryInterceptor(clientInt.UnaryClientInterceptor), grpc.WithStreamInterceptor(clientInt.StreamClientInterceptor))
	return NewResourceStoreClient(grpchan.InterceptClientConn(conn), nil
```

gamab replied:
Seems it was done like this before :shrug:

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
unistor client side updates

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
updated namespace extractor usage

gamab commented on pkg/storage/unified/resource/client.go:
```diff
@@ -102,6 +146,45 @@ func idTokenExtractor(ctx context.Context) (string, error) {
 	return "", fmt.Errorf("id-token not found")
 }
 
+func namespaceExtractor(ctx context.Context) (string, []string, error) {
```
I'm wondering if the namespace checker might be redundant given that we're retrieving the namespace from the user's identity. Currently, it seems like a no-op: we set the user's namespace in the header client-side, then verify it server-side against the user's identity.

The only scenario where the namespace header seems truly useful is if the caller's namespace is '*'. In that case, specifying the desired namespace in the header could be useful.

drclau replied:
cc: @ryantxu @kalleep

drclau replied:
Discussed live. The agreement is to keep the namespace checkers even though now they don't have a real utility, because there are future scenarios in which they'll be helpful.

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
Fixed integration tests

drclau committed:
introducing `mode` config for gRPC auth server & client side

DanCech commented on pkg/services/authn/grpcutils/config.go:
```diff
@@ -0,0 +1,66 @@
+	return &GrpcClientConfig{
+		Token:            section.Key("token").MustString(""),
+		TokenExchangeURL: section.Key("token_exchange_url").MustString(""),
+		TokenNamespace:   section.Key("token_namespace").MustString("stack-" + cfg.StackID),
```
stack or stacks?

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
Update pkg/services/authn/grpcutils/config.go

Co-authored-by: Dan Cech <dcech@grafana.com>

gamab committed:
PR Feedback

gamab commented on pkg/services/authn/grpcutils/config.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+const (
+	ModeOnPrem Mode = "on-prem"
```
We should probably align the modes in the authz package to these after

gamab replied:
I share your point of view.

gamab commented on pkg/services/authn/grpcutils/ns_authorizer.go:
```diff
@@ -0,0 +1,30 @@
+	if cfg.StackID != "" {
+		na = authzlib.NewNamespaceAccessChecker(
+			claims.CloudNamespaceFormatter,
+			authzlib.WithIDTokenNamespaceAccessCheckerOption(true),
```
If we want to allow service calls, we should probably set the `required` bool to `false`.

gamab commented on pkg/services/authz/config.go:
```diff
@@ -30,30 +29,35 @@ type Cfg struct {
-	tokenNamespace := section.Key("token_namespace").MustString("stack-" + cfg.StackID)
+	token := grpcClientAuthSection.Key("token").MustString("")
+	tokenExchangeURL := grpcClientAuthSection.Key("token_exchange_url").MustString("")
+	tokenNamespace := grpcClientAuthSection.Key("token_namespace").MustString("stack-" + cfg.StackID)
```
```suggestion
	tokenNamespace := grpcClientAuthSection.Key("token_namespace").MustString("stacks-" + cfg.StackID)
```

gamab committed:
Update pkg/services/authz/config.go

gamab commented on pkg/storage/unified/resource/client.go:
```diff
@@ -102,6 +150,45 @@ func idTokenExtractor(ctx context.Context) (string, error) {
+		return "", nil, err
+	}
+
+	namespace := caller.GetAllowedKubernetesNamespace()
```
I'm still puzzled by this function :thinking: 
How is it different from the caller's namespace?

drclau replied:
This code was removed from this PR. We'll deal with this question in a separate PR.

gamab commented on pkg/storage/unified/resource/client.go:
```diff
@@ -75,6 +73,56 @@ func NewLocalResourceClient(server ResourceServer) ResourceClient {
+		authnlib.WithMetadataExtractorOption(namespaceCloudExtractor(cfg.StackID)),
+	}
+
+	if cfg.Env == setting.Dev {
+		opts = allowInsecureTransportOpt(&grpcClientConfig, opts)
+	}
```
Should this also be set for the on-prem case.

gamab reviewed (approved):
LGTM. I tested the cloud mode in a docker-compose environment with all services at stake. It worked.

As we discussed together but to make it public: we'll have to do the conf changes on impacted services before merging this.

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

drclau committed:
added server side gRPC authn fallback-to-legacy mechanism

- brought back the old gRPC authenticator
- added `grpc_server_authentication.legacy_fallback` config option
- introduced `AuthenticatorWithFallback`
- added telemetry to track fallbacks

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
allow using the legacy resource client via

drclau committed:
remove the NamespaceAuthorizer

The NamespaceAuthorizer would fail in legacy mode. It will be added back in the future.

drclau committed:
use FlagAppPlatformGrpcClientAuth to enable new behavior, instead of legacy

drclau committed:
avoid double registration of metrics (fallbackCounter)

drclau committed:
remove client side namespace extractor

drclau committed:
use sync.Once as it's more idiomatic

drclau committed:
always enable FlagAppPlatformGrpcClientAuth for k8s int tests

drclau committed:
cleanup

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
imports cleanup

drclau committed:
switched to features.IsEnabledGlobally()

gamab reviewed (approved):
LGTM! I tested creating playlists on two instances, one using the legacy authn the other the new and it worked in both cases.

I went through the resource server code to make sure we would not break anything switching from a `StaticRequester` to `AuthInfo` . And I think this is safe. I'd like @ryantxu or @chaudyg to confirm :)

drclau committed:
Update pkg/services/authn/grpcutils/grpc_authenticator.go

Co-authored-by: Gabriel MABILLE <gamab@users.noreply.github.com>

drclau commented on pkg/services/authn/grpcutils/grpc_authenticator.go:
```diff
@@ -12,3 +63,75 @@ func NewInProcGrpcAuthenticator() *authnlib.GrpcAuthenticator {
+		return nil, err
+	}
+
+	authenticator, err := NewGrpcAuthenticator(cfg)
```
Sounds like a good simplification to me. Fixing...

drclau replied:
✅

drclau committed:
PR feedback: simplified fallback implementation

Co-Authored-By: Gabriel MABILLE <gamab@users.noreply.github.com>

ryantxu commented on pkg/services/authn/grpcutils/config.go:
```diff
@@ -0,0 +1,68 @@
+	SigningKeysURL   string
+	AllowedAudiences []string
+	Mode             Mode
+	LegacyFallback   bool
```
Is this just so we can stage the rollout?  EG, 
1. make sure this is deployed to all ST cloud instances with `legacy_fallback=true`
2. Update the MT unified storage
3. update all ST cloud instances with `legacy_fallback=false`
4. remove the fallback
?

drclau replied:
Just to add more clarify: the `legacy_fallback` flag is only used on the Unified Storage side. On Grafana side, there's a feature flag: `appPlatformGrpcClientAuth`.

* `legacy_fallback`: when enabled, if the new authenticator fails, the legacy authenticator is tried, and the result of the legacy authenticator decides the outcome.
* `appPlatformGrpcClientAuth`: when enabled, the new client implementation is used on Grafana side, which means an ID token and/or Access token will be added to the request metadata, which is then verified on UniStore side. When this is disabled, the old client will be used, which will pass some basic identity information as metadata to allow the reconstruction of the identity on UniStor side (as a `StaticRequester`).

When all instances of Grafana and UniStor are running the latest code, we will be able to remove the flags, the code added to support these flags, and the legacy code (old client, old unistor authenticator).

ryantxu commented on pkg/services/authn/grpcutils/config.go:
```diff
@@ -0,0 +1,68 @@
+		Token:            section.Key("token").MustString(""),
+		TokenExchangeURL: section.Key("token_exchange_url").MustString(""),
+		TokenNamespace:   section.Key("token_namespace").MustString("stacks-" + cfg.StackID),
+		Mode:             mode,
```
do we need mode?   is that the same as having a "remote_address" or a token_exchange_url that is not empty?

(weirdly this would be the first place we have a setting that explicitly says onprem vs cloud)

drclau replied:
Unfortunately we do, because for both cloud and on-prem setups the config options will have non-empty values. Now, we could potentially identify the type of setup based on the values of `remote_address` and `TokenExchangeURL`, but I feel that's hacky and non-obvious.

However, this will also go away as soon as we support Access tokens on-prem. This is really the main difference between the two setups.

ryantxu replied:
if it is just a temporary thing and only in cloud -- you can use the existence of `cfg.StackID` as an indication we are in "cloud mode" rather than adding a new setting

drclau replied:
hmm, yes, on client-side we can make use of `cfg.StackID`, if it's guaranteed that it's going to be empty for on-prem. Is this the case?

drclau replied:
@gamab, iirc, we debated this some time ago. Do you remember what the arguments against using `cfg.StackID` were?

drclau committed:
Merge branch 'main' into drclau/unistor/replace-authenticators-3

drclau committed:
conflict resolution

ryantxu commented on pkg/storage/unified/resource/client.go:
```diff
@@ -17,9 +19,13 @@ import (
 	"github.com/grafana/grafana/pkg/apimachinery/identity"
 	"github.com/grafana/grafana/pkg/services/auth"
 	"github.com/grafana/grafana/pkg/services/authn/grpcutils"
+	"github.com/grafana/grafana/pkg/setting"
```
hymmm -- I don't think this can be here (but also surprised the `pkg/services/auth` can!). This package has its own [go.mod](https://github.com/grafana/grafana/blob/main/pkg/storage/unified/resource/go.mod) so that we can build different implementations... see https://github.com/grafana/exp-unified-storage/blob/main/go.mod#L9

we should really add lint checking to make sure this fails!

ryantxu commented on pkg/storage/unified/resource/client.go:
```diff
@@ -17,9 +19,13 @@ import (
 	grpcUtils "github.com/grafana/grafana/pkg/storage/unified/resource/grpc"
 )
 
+// TODO(drclau): decide on the audience for the resource store
+const resourceStoreAudience = "resourceStore"
```
No strong opinion what it is called -- do any conventions exist yet?  "resource-store" or "unified-storage" make sense to me.

I do think everything in this client need to be moved up one folder -- into the other `client.go`.  In that case we will need to make the audience public so it can be referenced in both packages

drclau committed:
PR feedback: remove GrpcClientConfig.Mode, rely on cfg.StackID instead

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

drclau commented on pkg/storage/unified/resource/client.go:
I have attempted this, but we end up having import cycles. Do you mind if we first merge it as is, and move it in another PR?

drclau committed:
lint: use the same receiver name

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
Account for PR feedback

gamab committed:
Revert "Account for PR feedback"

This reverts commit 72af0a6680dd4f01464683a3f24e49d349ec8ff9.

gamab committed:
Proposal

gamab commented on pkg/services/authz/config.go:
```diff
@@ -30,30 +29,35 @@ type Cfg struct {
-	if mode == ModeCloud && token == "" && tokenExchangeURL == "" {
-		return nil, fmt.Errorf("authorization:  missing token or tokenExchangeUrl")
+	// When running in cloud mode, the token and tokenExchangeURL are required.
+	if cfg.StackID != "" && token == "" && tokenExchangeURL == "" {
```
```suggestion
	if mode=ModeGRPC && cfg.StackID != "" && token == "" && tokenExchangeURL == "" {
```

gamab committed:
Update pkg/services/authz/config.go

gamab commented on pkg/services/authz/config.go:
```diff
@@ -30,30 +29,35 @@ type Cfg struct {
-	if mode == ModeCloud && token == "" && tokenExchangeURL == "" {
-		return nil, fmt.Errorf("authorization:  missing token or tokenExchangeUrl")
+	// When running in cloud mode, the token and tokenExchangeURL are required.
+	if mode==ModeGRPC && cfg.StackID != "" && token == "" && tokenExchangeURL == "" {
```
```suggestion
	if mode == ModeGRPC && cfg.StackID != "" && token == "" && tokenExchangeURL == "" {
```

gamab committed:
Update pkg/services/authz/config.go

gamab commented on pkg/services/authz/config.go:
```diff
@@ -30,30 +29,35 @@ type Cfg struct {
-	if mode == ModeCloud && token == "" && tokenExchangeURL == "" {
-		return nil, fmt.Errorf("authorization:  missing token or tokenExchangeUrl")
+	// When running in cloud mode, the token and tokenExchangeURL are required.
+	if mode == ModeGRPC && cfg.StackID != "" && token == "" && tokenExchangeURL == "" {
+		return nil, fmt.Errorf("authorization: missing token or tokenExchangeUrl")
```
```suggestion
	if mode == ModeGRPC && cfg.StackID != "" {
		if token == "" || tokenExchangeURL == "" {
			return nil, fmt.Errorf("authorization:  missing token or tokenExchangeUrl")
		}
```

gamab committed:
Update pkg/services/authz/config.go

gamab committed:
extracted changes in #95120

gamab commented:
/deploy-to-hg

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
extracted server side changes in #95086

gamab committed:
Comment out feature toggle while server side is not merged

gamab commented:
/deploy-to-hg

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab committed:
No error needed

gamab committed:
Remove ToDo

gamab committed:
Remove ToDo

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab commented:
/deploy-to-hg

gamab committed:
Merge remote-tracking branch 'origin/main' into drclau/unistor/replace-authenticators-3

gamab commented:
/deploy-to-hg
